### PR TITLE
Store multiple IPs in keychains

### PR DESCRIPTION
**Pull request by LinnierGames**
### What has changed
- keychain at the IP key stores a list of strings vs one string

### What's new
- Created UserPreferences to manage reading and writing to keychains

### CODE REVIEW

**Review by LinnierGames — COMMENTED** [review]
Self reviewed

File: PicPi/AddFrameViewController.swift [hunk 1]
```diff
@@ -13,6 +13,8 @@ class AddFrameViewController: UIViewController, UITextFieldDelegate  {
   let ipTextField = customUITextFieldWithDecimalPad()
   var margins: UILayoutGuide!
   let mqttManager =  MQTTManager()
+  private let userPreferences = injectUserPreferences()
```
Comment (RawiSoft) [inline PicPi/AddFrameViewController.swift:16]: Property Injection ?

File: PicPi/HomeViewController.swift [hunk 2]
```diff
@@ -26,14 +25,9 @@ class HomeViewController: UIViewController     {
   override func viewWillAppear(_ animated: Bool) {
     /// try to make it simple for now by updating the table if there is any saved IP Address
     super.viewWillAppear(animated)
-    if let ipData = KeyChain.loadKey(Constants.KeyChains.piIPAddressKey) {
-      ip = String(decoding: ipData, as: UTF8.self)
-      searchButton.isHidden = true
-    }
-    
+    addButton.isHidden = userPreferences.ipAddresses().isEmpty
```
Comment (RawiSoft) [inline PicPi/HomeViewController.swift:28]: Nice!!

File: PicPi/UserPreferencesImpl.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import Foundation
+
+func injectUserPreferences() -> UserPreferences {
+  return UserPreferencesImpl()
+}
+
+private class UserPreferencesImpl: UserPreferences {
+  private let piIPAddressKey = "Pi1IP"
+  private lazy var encoder = JSONEncoder()
+  private lazy var decoder = JSONDecoder()
```
Comment (RawiSoft) [inline PicPi/UserPreferencesImpl.swift:17]: JSON to decode and encode Array of IPs?

**Review by RawiSoft — APPROVED** [review]

File: PicPi/UserPreferencesImpl.swift (hunk 3 above)
Comment (LinnierGames) [reply PicPi/UserPreferencesImpl.swift:17]: Yeah I was thinking we might start to store more than an array of strings like name, so might as well start using json

wdyt?

File: PicPi/AddFrameViewController.swift (hunk 1 above)
Comment (LinnierGames) [reply PicPi/AddFrameViewController.swift:16]: Something like it lol so view controllers don't get tested like business logic types so injecting it here vs through the init is okay

If the inject method needs arguments (eg `injectFrameAPI(host: myPIURL)`) then the service might need to be injected from the init

**Comment by LinnierGames** [comment]
[Update keychains to store multiple IPs](https://trello.com/c/XBPW4dcl/55-update-keychains-to-store-multiple-ips)

**Pull request merged**